### DISCUSSION TITLE
How to implicitly wait for database creation condition before executing database queries

### DISCUSSION THREAD

**Discussion by Ahsoka** [sqlalchemy/sqlalchemy#7883] [Usage Questions] [answered]
Hello!  I was wondering if there was a feature in SQLAlchemy to prevent a database transaction until the database is actually created.

Consider the following example:
```py
from sqlalchemy.ext.asyncio import AsyncSession
from .tables import Thing, mapper
# NOTE: Thing declarative mapping and declarative base are defined in another file

import math
import aiohttp.web

app = aiohttp.web.Application()
routes = aiohttp.web.RouteTableDef()

engine = ...

@routes.get('/create-db')
async def create_db(request: aiohttp.web.Request):
    # Something super long and time consuming here
    await some_code_that_takes_along_time()
    
    async with engine.begin() as conn:
        await conn.run_sync(mapper.metadata.create_all)

@routes.get('/create-some-object')
async def create_some_object(request: aiohttp.web.Request):
    async with AsyncSession(engine) as session:
        thing = Thing(number=math.tau, params=request.params)
        session.add(thing)
        
        # NOTE: Now wait until the database is created until trying
        # to add the object the database.
```

If someone accesses the `/create_some_object` endpoint before the `/create_db` I want whatever the object is passed into the `/create_some_object` endpoint to be created at some point (ideally right after the database is created).  Is there a way to do this in SQLAlchemy or do I need to create a custom implementation for this?

A possible solution to my problem is to use an `asyncio.Event` which would signify when the database has been created, however, in my code I have a ton of endpoints so it would be pain to add `await event.wait()` to each endpoint which is the main motivation to handle this _implicitly_.

**Comment by CaselIT** [comment]
Hi,

Off the top of my head I don't think there is anything that does this in sqlalchemy. It seems to be a very application specific requirement.
Usually web servers let you define a middleware or similar functionality that can run before the request function code. In that middleware you could check if the db exists and act accordingly (like returning an error to the caller).

**Comment by zzzeek** [comment, marked as answer]
creating tables in the scope of a web request is not a normal thing to do and we would assume this is a very special call within some administrative application.    it would not be a normal thing for a web application to be accepting normal "do thing" types of requests while administrative activity is going on.

that said, if you want one web request to defer the thing it produces, you first need to decide if you want the operation to block or not.  if you want to block, use an asyncio.Lock (very dangerous to do this in a web app).  If you want it to return immediately and the thing gets done eventually, use a background task queue.
